### ISSUE TITLE
Add ignoreJSX option to no-undef

### ISSUE THREAD

**Issue by dmin** [eslint/eslint#10754]
**What rule do you want to change?**
no-undef

**Does this change cause the rule to produce more or fewer warnings?**
Fewer (If proposed new option is enabled)

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option, `ignoreJSX`, to optionally allow ignoring undeclared variables in JSX expressions.

Valid values for the new option would be `true` and `false`.

**Please provide some example code that this change will affect:**

The following would produce no errors:
```js
/*eslint no-undef: ["error", { "ignoreJSX": true }] */
var template = <div>{anUndeclaredVariable}</div>;

```

**What does the rule currently do for this code?**
The following would produce an error:
```js
/*eslint no-undef: ["error"] */
var template = <div>{anUndeclaredVariable}</div>;
```

**What will the rule do after it's changed?**
The `no-undef` rule will optionally ignore undeclared variables in JSX expressions if the `ignoreJSX` option is set to `true`. The motivation for this change is to enable writing AngularJS templates with standard JSX.

AngularJS templates are evaluated in the context of a "scope" at run time, so we're unable to statically lint for undeclared variables in templates when written in JSX.

```js
// AngularJS template as HTML string
var template = '<div>{{myVariable}}</div>'

// AngularJS template as JSX (transformed to the above HTML string at build time)
var template = <div>{myVariable}</div>
```

We can obviously write a custom rule, but it would essentially be a fork of the built-in `no-undef` rule, so wanted to check if this new option would be useful enough to others to warrant an addition to the built-in rule.

I'd be happy to open a PR implementing this change.
